Sources/Action.swift

```diff
@@ -109,21 +95,21 @@ public final class Action<Input, Output, Error: Swift.Error> {
 	///            producer.
 	public func apply(_ input: Input) -> SignalProducer<Output, ActionError<Error>> {
 		return SignalProducer { observer, disposable in
-			var startedExecuting = false
+			var executedProducer: SignalProducer<Output, Error>?
 
-			self.executingQueue.sync {
-				if self._isEnabled.value {
-					self._isExecuting.value = true
-					startedExecuting = true
+			self._state.modify { state in
+				if state.isEnabled {
+					state.isExecuting = true
+					executedProducer = self.executeClosure(input)
```

Comment:
The reason we started on this is so that this:

```swift
	public init<P: PropertyProtocol, T>(input: P, _ execute: @escaping (T) -> SignalProducer<Output, Error>) where P.Value == T? {
		self.init(enabledIf: input.map { $0 != nil }) {
			execute(input.value!)
		}
	}
```

Could guarantee that `input.value` was non-nil. I actually don't think the code as written solves even this.